src/acceleration/BaseQNAcceleration.cpp

```diff
@@ -666,5 +581,241 @@ void BaseQNAcceleration::writeInfo(
   }
   _infostringstream << std::flush;
 }
+
+void BaseQNAcceleration::concatenateCouplingData(
+    const DataMap &cplData, const std::vector<DataID> &dataIDs, const std::vector<DataID> &primaryDataIDs, Eigen::VectorXd &values,
+    Eigen::VectorXd &oldValues, Eigen::VectorXd &primaryValues, Eigen::VectorXd &oldPrimaryValues)
+{
+
+  // Needs to be called every iteration, since the time window size can vary with participant first
+  moveTimeGridToNewWindow(cplData, dataIDs);
+
+  /// If not reduced quasi-Newton then sample the residual of data in dataIDs to the corresponding time grid in _timeGrids and concatenate everything into a long vector
+  if (!_reduced) {
+    Eigen::Index offset = 0;
+
+    for (int id : _primaryDataIDs) {
+      auto            waveform = cplData.at(id)->timeStepsStorage();
+      Eigen::Index    dataSize = cplData.at(id)->values().size();
+      Eigen::VectorXd timeGrid = _timeGrids.at(id);
+
+      for (int i = 0; i < timeGrid.size(); i++) {
+
+        Eigen::VectorXd primaryData    = waveform.sample(timeGrid(i));
+        Eigen::VectorXd oldPrimaryData = cplData.at(id)->getPreviousValuesAtTime(timeGrid(i));
+
+        PRECICE_ASSERT(primaryValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");
+        PRECICE_ASSERT(oldPrimaryValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");
+
+        for (Eigen::Index i = 0; i < dataSize; i++) {
+          primaryValues(i + offset)    = primaryData(i);
+          oldPrimaryValues(i + offset) = oldPrimaryData(i);
+        }
+        offset += dataSize;
+      }
+    }
+  } else {
+    Eigen::Index offset = 0;
+    for (auto id : _primaryDataIDs) {
+      Eigen::Index dataSize       = cplData.at(id)->values().size();
+      auto &       primaryData    = cplData.at(id)->values();
+      const auto & oldPrimaryData = cplData.at(id)->previousIteration();
+
+      PRECICE_ASSERT(primaryValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");
+      PRECICE_ASSERT(oldPrimaryValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");
+
+      for (Eigen::Index i = 0; i < dataSize; i++) {
+        primaryValues(i + offset)    = primaryData(i);
+        oldPrimaryValues(i + offset) = oldPrimaryData(i);
+      }
+      offset += dataSize;
+    }
+  }
```

Comment:
I'm not so happy with the amount of code duplication here and at other points where you are branching depending on `_reduced`. Why not do the following:

```
Eigen::Index offset = 0;
for (int id : _primaryDataIDs) {
  Eigen::VectorXd timeGrid;
  if (!_reduced) {
    timeGrid = _timeGrids.at(id);
  } else {
    timeGrid = ...; // reduced time grid only containing _timeGrids.at(id).last();
  }

  // do everything like implemented in the "if (!_reduced)"-part; should be possible to generalize the "else" part correspondingly.
}
```

This also has the side effect that you would use `waveform.sample(timeGrid(i));` instead of `cplData.at(id)->values()`. Sampling from `waveform` is in my opinion preferrable since `values()` corresponds to the raw data and is a bit more low-level.

Please also use `CouplingData::getSize()` and not `values().size()`. Reason: at some point it would be nice to completely get rid off `values()` and only access the `CouplingData::timeStepsStorage()` (i.e. the waveform).